### PR TITLE
[ZP-36] Close interpreters in multithreaded mode

### PR DESCRIPTION
**Pull request by egorklimov**
### What is this PR for?

This PR provides multithreading realization of interpreters closing.

Benchmark based on [JMH](http://openjdk.java.net/projects/code-tools/jmh/) - `ZeppelinServerBenchmarkTest`:
* Creates `InterpreterSettingManager` and `InterpreterFactory`;
* Runs 6 interpreters (`python`, `sh`, `md`, `groovy`, `angular`, `spark`) for each user in isolated mode;
* Measures InterpreterSettingManager closing.

Result with usersCnt - 7, 5, 3, 1 (42, 30, 18, 6 interpreter processes):
```
Benchmark                                                 (usersCnt)  Mode  Cnt   Score   Error  Units
ZeppelinServerBenchmarkTest.testDefaultServerShutdown              7  avgt   10  16,690 ± 0,191   s/op
ZeppelinServerBenchmarkTest.testDefaultServerShutdown              5  avgt   10  11,532 ± 0,064   s/op
ZeppelinServerBenchmarkTest.testDefaultServerShutdown              3  avgt   10   6,942 ± 0,020   s/op
ZeppelinServerBenchmarkTest.testDefaultServerShutdown              1  avgt   10   2,348 ± 0,018   s/op
ZeppelinServerBenchmarkTest.testServerShutdownInParallel           7  avgt   10   3,612 ± 0,317   s/op
ZeppelinServerBenchmarkTest.testServerShutdownInParallel           5  avgt   10   2,635 ± 0,075   s/op
ZeppelinServerBenchmarkTest.testServerShutdownInParallel           3  avgt   10   2,473 ± 0,072   s/op
ZeppelinServerBenchmarkTest.testServerShutdownInParallel           1  avgt   10   2,378 ± 0,029   s/op
```
[benchmark.zip](https://github.com/TinkoffCreditSystems/zeppelin/files/2591694/benchmark.zip) contains: 
* `Benchmark.patch` - patch with `ZeppelinServerBenchmarkTest`;
* `benchmark.csv` - benchmark result;
jmh-benchmark
* `benchmark-result.json` - results (you could visualize it using [JMH Visualizer](http://jmh.morethan.io/))
zp-36

### How to run benchmark 
1. Apply patch from `benchmark.zip`
2. Add [IntelliJ IDEA JMH Plugin](https://plugins.jetbrains.com/plugin/7529-jmh-plugin)
3. Build Zeppelin 
4. Run ZeppelinServerBenchmarkTest in IntelliJ IDEA

### What type of PR is it?
Refactoring

### What is the Jira issue?
ZP-36
https://github.com/apache/zeppelin/pull/3229

### How should this be tested?
* CI pass

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSetting.java [hunk 1]
```diff
@@ -477,10 +478,20 @@ void closeInterpreters(String user, String noteId) {
 
   public void close() {
     LOGGER.info("Close InterpreterSetting: " + name);
-    for (ManagedInterpreterGroup intpGroup : interpreterGroups.values()) {
-      intpGroup.close();
-    }
+    List<Thread> closeThreads = interpreterGroups.values().stream()
+            .map(g -> new Thread(g::close, name + "-close"))
+            .peek(Thread::start)
```
Comment (aka-demik) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSetting.java:485]: Предлагаю добавить потокам UncaughtExceptionHandler с логгером.
```suggestion
            .peek(t -> t.setUncaughtExceptionHandler((th, e) ->
                LOGGER.error("InterpreterSetting close error", e)))
            .peek(Thread::start)
```

Мотивация простая, если в одном из `close()` вылетал не обработанный `Exception`, то раньше он подымался до самого [main()](https://github.com/TinkoffCreditSystems/zeppelin/blob/becfdf15f6cc80669fe70ae4443a21c807946827/zeppelin-server/src/main/java/org/apache/zeppelin/server/ZeppelinServer.java#L289-L298) и там попадал в лог. А в новой версии можем его потерять...

**Review by aka-demik — CHANGES_REQUESTED** [review]

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSetting.java (hunk 1 above)
Comment (egorklimov) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSetting.java:485]: Поправил в 4bab30f76f302b30227d62377cd735bf1c723896.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/ManagedInterpreterGroup.java [hunk 2]
```diff
@@ -118,27 +131,44 @@ private void close(Collection<Interpreter> interpreters) {
...
+            .peek(Thread::start)
+            .collect(Collectors.toList());
+
+    for (Thread t : closeThreads) {
       try {
-        interpreter.close();
-      } catch (InterpreterException e) {
-        LOGGER.warn("Fail to close interpreter " + interpreter.getClassName(), e);
-      }
-      //TODO(zjffdu) move the close of schedule to Interpreter
```
Comment (aka-demik) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/ManagedInterpreterGroup.java:135]: Предлагаю todo сохранить. Мысль в ней дельная... Просто сдвинем в строку 169

Comment (egorklimov) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/ManagedInterpreterGroup.java:135]: Поправил в e7abdd305df86595c2c11f4743e7d9cbb0558206

**Review by aka-demik — APPROVED** [review]

**Pull request merged**